Add suites directory

hwikle-lanl opened the pull request:
Code review checklist:

- [ ] Code is generally sensical and well commented
- [ ] Variable/function names all telegraph their purpose and contents
- [ ] Functions/classes have useful doc strings
- [ ] Function arguments are typed
- [ ] Added/modified unit tests to cover changes.
- [ ] New features have documentation added to the docs.
- [ ] New features and backwards compatibility breaks are noted in the RELEASE.md

hwikle-lanl committed:
Add suites directory

Paul-Ferrell commented:
I'm pretty sure you're not actually ready for me to review this, but I see a few issues already.
The expected directory structure is:
`suites/<suite_name>/`
So for whatever suite you're looking at, you should be finding files under the specific directory for a suite rather than the general `suites` directory.

hwikle-lanl committed:
Progress towards solution

hwikle-lanl committed:
Further progress

hwikle-lanl committed:
Fix style issues

hwikle-lanl committed:
Increase timeout for test_owner

Test had been passing on develop, but no longer is. Increasing
the timeout from 1 to 2 seconds causes it to pass. Something
has probable changed that is slowing the filesystem down.

hwikle-lanl committed:
Add suites tests

hwikle-lanl committed:
Fix mistake

hwikle-lanl committed:
Make resolver check both tests and suites dirs

hwikle-lanl committed:
Improve suites test

hwikle-lanl committed:
Pass basic suites test

hwikle-lanl committed:
Fix regression

hwikle-lanl committed:
Add bare YAML suites test

hwikle-lanl committed:
Stub out additional suites tests

hwikle-lanl committed:
Refactor find_config

hwikle-lanl committed:
Miscellaneous changes

hwikle-lanl committed:
Further progress

hwikle-lanl committed:
Fix regression

hwikle-lanl committed:
Progress towards proper parsing of host config

hwikle-lanl committed:
Progress towards solution

hwikle-lanl committed:
Further progress towards solution

hwikle-lanl committed:
Further progress

hwikle-lanl committed:
Reorganize resolver code

hwikle-lanl committed:
Rename tests appropriately

hwikle-lanl committed:
Flesh out unit tests

hwikle-lanl committed:
Implement hosts from suites dir

hwikle-lanl committed:
Implement os config from suites dir

hwikle-lanl committed:
Implement loading mode from suites dir

hwikle-lanl committed:
Add timeouts to unit tests to prevent hanging

hwikle-lanl committed:
Fix regression in series_config

hwikle-lanl committed:
Add missed timeout

hwikle-lanl committed:
Fix error in safe_load_config

hwikle-lanl committed:
Fix date math (again)

hwikle-lanl committed:
Rename func_utils to micro

hwikle-lanl committed:
Add return type

hwikle-lanl committed:
Flesh out hash test

hwikle-lanl committed:
Fix error due to commented line

hwikle-lanl committed:
Resolve a bunch of failing tests (WOOHOO!!)

hwikle-lanl committed:
Remove legacy data

hwikle-lanl committed:
Revert "Remove legacy data"

This reverts commit 2c7f6fab294451a5b8786501fa6ea6b3387bfa04.

hwikle-lanl committed:
Fix error due to introduction of ConfigInfo

hwikle-lanl committed:
Implement true directory hashing

hwikle-lanl committed:
Improve directory hashing algorithm

hwikle-lanl committed:
Add missing import

hwikle-lanl committed:
Ignore yaml config files when hashing directories

hwikle-lanl committed:
Fix style issues

hwikle-lanl committed:
Remove stray pdb statement

hwikle-lanl committed:
Pull parsetime bugfix over from other branch

hwikle-lanl committed:
Pass hashing unit test

hwikle-lanl committed:
Remove unused test data

hwikle-lanl committed:
Refactor find_file

hwikle-lanl committed:
Increase timeout for series cmd tests

hwikle-lanl committed:
Fix bug in find_file

hwikle-lanl committed:
Fix suite path (hopefully)

hwikle-lanl committed:
Fix source path issue (finger crossed)

hwikle-lanl committed:
Fix download destination

Paul-Ferrell commented on lib/pavilion/filters/parse_time.py:
```diff
@@ -137,3 +142,24 @@ def normalize(unit: str) -> str:
     return unit
+
+
+def safe_update(date: datetime,
+                year: int = None,
+                month: int = None, day: int = None) -> datetime:
```
Docstring needed.

Paul-Ferrell commented on lib/pavilion/filters/parse_time.py:
```diff
@@ -62,12 +63,16 @@ def parse_duration(rval: str, now: datetime) -> datetime:
         raise ValueError(f"Invalid unit {unit} for duration")
 
     if unit == 'years':
-        return now.replace(year=now.year - mag)
+        return safe_update(now, year=now.year - mag)
 
```
It feels like using datetime.timedelta would be good enough, but this is better.

hwikle-lanl replied:
`timedelta` unfortunately can't handle variable-size units like months and years, which is what makes this necessary.

Paul-Ferrell commented on lib/pavilion/commands/config.py:
```diff
@@ -244,7 +244,8 @@ def create_config_dir(self, pav_cfg: config.PavConfig, path: Path,
                                  .format(config_file_path), err)
 
-        for subdir in 'hosts', 'modes', 'tests', 'os', 'test_src', 'plugins', 'collections':
+        for subdir in ('hosts', 'modes', 'tests', 'os', 'test_src',
+            'plugins', 'collections', 'suites'):
```
Since we're replacing test_src and tests with suites, let's get rid of those here.

hwikle-lanl replied:
Since we're technically deprecating those, would it make sense to keep them around for awhile for backwards compatibility? Otherwise we'll have to restructure all the existing tests as soon as this is rolled out.

Paul-Ferrell commented on lib/pavilion/builder.py:
```diff
@@ -965,17 +978,33 @@ def _hash_io(cls, contents):
+        """Recursively hash the files in the given directory, excluding
+        files with the given suffixes.
+
+        We might need to modify this to handle circular symlinks.
```
We definitely need to handle this.

Paul-Ferrell replied:
Or, just don't follow symlinks (simply add their pointer dest to the hash).

Paul-Ferrell commented on lib/pavilion/series/test_set.py:
```diff
@@ -722,14 +725,22 @@ def wait(self, wait_for_all=False, wait_period: int = TEST_WAIT_PERIOD) -> int:
 
         completed_tests = self.mark_completed()
 
+        start = time.monotonic()
+
```
Using the monotonic clock really seems like over kill.  I'd like to hear why it matters.

hwikle-lanl replied:
The choice was motivated by this note in the docs for `time.time`:

> Note that even though the time is always returned as a floating point number, not all systems provide time with a better precision than 1 second. While this function normally returns non-decreasing values, it can return a lower value than a previous call if the system clock has been set back between the two calls.

Paul-Ferrell commented on lib/pavilion/resolver/resolver.py:
```diff
@@ -636,7 +735,9 @@ def _load_raw_configs(self, request: TestRequest, modes: List[str],
 
             # Apply modes.
             try:
-                test_cfg = self.apply_modes(test_cfg, modes)
+                test_cfg = self.apply_modes(test_cfg, modes, request.suite)
+                test_cfg = self.apply_host(test_cfg, self._host, request.suite)
+                test_cfg = self.apply_os(test_cfg, self._os, request.suite)
```
The stack should be:
 - modes
 - test_layers
 - host
 - os
 - <empty config>

What you have is very backwards.

Paul-Ferrell replied:
IE settings in `host` should override `os`, and so on.

hwikle-lanl committed:
Add docstring for safe_update

hwikle-lanl committed:
Miscellaneous changes

hwikle-lanl committed:
Pass suite_with_source test

hwikle-lanl committed:
Fix regressions

hwikle-lanl committed:
Pass suite_with_source unittest

hwikle-lanl committed:
Pass circular_symlink unittest

hwikle-lanl committed:
Finish implementing PR feedback

hwikle-lanl committed:
Fix failing pylint tests in CI

hwikle-lanl committed:
Second try

hwikle-lanl committed:
Fix newly failing style tests

hwikle-lanl committed:
Update workflow file

hwikle-lanl committed:
Try something else

hwikle-lanl committed:
Fix failing status_history unittest (hopefully)

hwikle-lanl committed:
Try again to fix status_history unittest in CI

hwikle-lanl committed:
Attempt to fix demo unittest in CI

hwikle-lanl committed:
Try again

hwikle-lanl commented:
Merging, since all feedback from @Paul-Ferrell  is implemented.
